controlplane/kubeadm/internal/controllers/helpers.go

```diff
@@ -289,48 +292,140 @@ func (r *KubeadmControlPlaneReconciler) generateKubeadmConfig(ctx context.Contex
+	//		ClusterName:       cluster.Name,
+	//		Version:           &kcp.Spec.Version,
+	//		InfrastructureRef: *infraRef,
+	//		Bootstrap: clusterv1.Bootstrap{
+	//			ConfigRef: bootstrapRef,
+	//		},
+	//		FailureDomain:    failureDomain,
+	//		NodeDrainTimeout: kcp.Spec.MachineTemplate.NodeDrainTimeout,
+	//	},
+	//}
+	//if kcp.Spec.MachineTemplate.NodeDeletionTimeout != nil {
+	//	machine.Spec.NodeDeletionTimeout = kcp.Spec.MachineTemplate.NodeDeletionTimeout
+	//}
+	//
+	//// Machine's bootstrap config may be missing ClusterConfiguration if it is not the first machine in the control plane.
+	//// We store ClusterConfiguration as annotation here to detect any changes in KCP ClusterConfiguration and rollout the machine if any.
+	//clusterConfig, err := json.Marshal(kcp.Spec.KubeadmConfigSpec.ClusterConfiguration)
+	//if err != nil {
+	//	return errors.Wrap(err, "failed to marshal cluster configuration")
+	//}
+	//
+	//// Add the annotations from the MachineTemplate.
+	//// Note: we intentionally don't use the map directly to ensure we don't modify the map in KCP.
+	//for k, v := range kcp.Spec.MachineTemplate.ObjectMeta.Annotations {
+	//	machine.Annotations[k] = v
+	//}
+	//machine.Annotations[controlplanev1.KubeadmClusterConfigurationAnnotation] = string(clusterConfig)
+	//
+	//if err := r.Client.Create(ctx, machine); err != nil {
+	//	return errors.Wrap(err, "failed to create machine")
+	//}
+	//return nil
+
+	machine, err := r.computeDesiredMachine(ctx, kcp, cluster, infraRef, bootstrapRef, failureDomain, nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to compute desired machine")
 	}
-	if kcp.Spec.MachineTemplate.NodeDeletionTimeout != nil {
-		machine.Spec.NodeDeletionTimeout = kcp.Spec.MachineTemplate.NodeDeletionTimeout
+	patchOptions := []client.PatchOption{
+		client.ForceOwnership,
+		client.FieldOwner(kcpManagerName),
 	}
+	if err := r.Client.Patch(ctx, machine, client.Apply, patchOptions...); err != nil {
+		return errors.Wrap(err, "failed to create machine")
+	}
+	return nil
+}
 
-	// Machine's bootstrap config may be missing ClusterConfiguration if it is not the first machine in the control plane.
-	// We store ClusterConfiguration as annotation here to detect any changes in KCP ClusterConfiguration and rollout the machine if any.
-	clusterConfig, err := json.Marshal(kcp.Spec.KubeadmConfigSpec.ClusterConfiguration)
-	if err != nil {
-		return errors.Wrap(err, "failed to marshal cluster configuration")
+func (r *KubeadmControlPlaneReconciler) computeDesiredMachine(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, cluster *clusterv1.Cluster, infraRef, bootstrapRef *corev1.ObjectReference, failureDomain *string, existingMachine *clusterv1.Machine) (*clusterv1.Machine, error) {
+	desiredMachine := &clusterv1.Machine{}
+
+	var machineName string
+	var machineUID types.UID
+	var machineFailureDomain *string
+	var version *string
+	annotations := map[string]string{}
+	if existingMachine == nil {
```

Comment:
Similar to the other PR. I think it would be easier to just construct the whole Machine with everything that is equal and then modify it later with if/else depending on the machine already exists

(also the diff becomes a lot smaller)